src/main/java/com/google/sps/algorithms/AbuseDetection.java

```diff
@@ -0,0 +1,91 @@
+package com.google.sps.algorithms;
+
+import com.google.sps.data.MatchRepository;
+import com.google.sps.data.MatchRequest;
+import com.google.sps.data.PersistentUserRepository;
+import com.google.sps.data.User;
+import com.google.sps.data.UserRepository;
+import java.lang.Exception;
+import java.io.IOException; 
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+
+
+public final class AbuseDetection {
+
+    // private Collection <LocalTime> timesOfRequests = new ArrayList<LocalTime>();
+    private Collection <Date> timesOfRequests = new ArrayList<Date>();
+
+    private int requestCounter = 0;
+    private int requestsDropped = 0;
+
+    private final int currentNumRequestsAllowed;
+    private final Duration timePeriod;
+  
+  /**
+  * Constructor that initializes currentNumRequestsAllowed and timePeriod
+  * The durationInSeconds represents the amount of seconds passed 
+  * in for this time interval of requests.
+  * The requestsAllowed is the amount of requests that are being allowed
+  * during that timeValue interval.
+  */
+  public AbuseDetection(Duration requestInterval, int requestsAllowed) { 
+    this.timePeriod = requestInterval;
+    this.currentNumRequestsAllowed = requestsAllowed;
+  }
+
+  /**
+  *  Function takes a local time, tries to add requests only if there are not
+  *  more than the currentRequestsAllowed variable, and less than the 
+  *  timePeriod. If added the method returns true, if not it returns false.
+  */
+   public boolean addRequest(LocalTime currentTime) {
+        Instant currentTimeToInstance =  currentTime.atDate(LocalDate.now()).
+        atZone(ZoneId.systemDefault()).toInstant();
+        Date timeToDate = Date.from(currentTimeToInstance);
+    
+        boolean returnValue = false;
+        if(requestCounter < currentNumRequestsAllowed) {
+            timesOfRequests.add(timeToDate);
+
+            returnValue = true;
+            requestCounter++;
+
+        }
+        else {
+
+            for(Date currentDate : timesOfRequests) {
```

Comment:
It looks like your logic checks if every stored request is older than timePeriod, and if it finds one, it adds the new request and returns. Is this what you intended?

Also, could you add a few one-liner comments in this method explaining your strategy? Doesn't have to be much, just enough to document what blocks of the code are responsible for the 2 or 3 high-level tasks you're executing here (for instance, it wasn't clear to me as a reader that checking if timeDifference.compareTo(timePeriod) > 0 was checking that time timeDifference > timePeriod without reading the java docs.

Reply:
Yes, that's what I intended.